### PR TITLE
[CL-18] toast component and service

### PR DESCRIPTION
**Pull request by willmartian**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Adds a toast component and service to the CL. Remove direct client references to `ngx-toastr`

## Code changes

- **apps** 
  - Removed toast css from all clients and consolidated in CL
  - Replaced `ToastrService` with `ToastService` in app components
  - **messages.json:** Added variant name so SR can announce it upon opening 
- **libs/components** 
  - **toast.component:** Presentational component
  - **toastr.component:** `ngx-toastr` wrapper for `bit-toast`
  - **toast.service:** Facade over `ToastrService`
    - removed functionality to use custom HTML in toasts
- **libs/common**
  - deprecated `PlatfromUtilsService.showToast`

## Screenshots

See new stories in Storybook. 

image

image

https://github.com/bitwarden/clients/assets/17113462/86f842cb-2a54-4c0c-9768-6b04164bf56d

https://github.com/bitwarden/clients/assets/17113462/583bfb2c-4320-48d6-980b-6ebcb7b2000e

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by danielleflinn — DISMISSED** [review]
Looks good! I couldn't get the Service stories to work on my end, but I assume that is a bug with chromatic not storybook.

File: libs/components/src/toast/toastr.component.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { animate, state, style, transition, trigger } from "@angular/animations";
+import { Component } from "@angular/core";
+import { Toast as BaseToastrComponent } from "ngx-toastr";
+
+@Component({
+  selector: "[toast-component2]",
```
Comment (Hinton) [inline libs/components/src/toast/toastr.component.ts:6]: Why a `2`?

Comment (willmartian) [reply libs/components/src/toast/toastr.component.ts:6]: That what `ngx-toastr` expects. This component was already present, moved from [here.](https://github.com/bitwarden/clients/pull/6490/files#diff-7914c620b5900cf916dca16508529205bf5b551a5f9d1618349018722b04a655L14-L15) (I should have moved it in a separate commit so Git would recognize it as a move.) My main change was using `bit-toast` in the template.

File: libs/components/src/toast/toast.component.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  @Input({ required: true })
+  message: string | string[];
+
+  /** An optional title to display over the message. */
+  @Input() title: string;
+
+  /**
+   * The percent width of the progress bar, from 0-100
+   **/
+  @Input() progressWidth = 0;
```
[the remark this answers is not in the document]
Comment (willmartian) [reply libs/components/src/toast/toast.component.ts:50]: Inputs are supplied by the `toastr` [component](https://github.com/bitwarden/clients/pull/6490/files/07c5a515af0e5bccef4255e657cae846fb948840#diff-2a855d1135d1790bd13afd87a9fc12836ea895734fdc120bed2d3fda809329a8R8-R14)

That component extends a base component that the library provides. 

I wanted `bit-toast` to be completely presentation and separate from the library.

File: libs/components/src/toast/toastr.component.ts (hunk 1 above)
Comment (Hinton) [reply libs/components/src/toast/toastr.component.ts:6]: The demo of using a pink component doesn't require specific selector https://github.com/scttcper/ngx-toastr#using-a-custom-toast

Comment (willmartian) [reply libs/components/src/toast/toastr.component.ts:6]: Huh, you are right. Don't know why I thought otherwise. I removed the selector entirely since this only presented by a controller.

File: libs/common/src/platform/abstractions/platform-utils.service.ts [hunk 3]
```diff
@@ -28,6 +28,7 @@ export abstract class PlatformUtilsService {
   getApplicationVersionNumber: () => Promise<string>;
   supportsWebAuthn: (win: Window) => boolean;
   supportsDuo: () => boolean;
+  /** @deprecated use `@bitwarden/components/ToastService.showToast` instead */
```
[the remark this answers is not in the document]
Comment (willmartian) [reply libs/common/src/platform/abstractions/platform-utils.service.ts]: Added epic link: https://github.com/bitwarden/clients/pull/6490/commits/fb381830dc923dca5d498fe4323676b1d7f71391

File: libs/components/src/toast/toast.component.html [hunk 4]
```diff
@@ -0,0 +1,22 @@
+<div
+  class="tw-mb-1 tw-min-w-[300px] tw-text-contrast tw-flex tw-flex-col tw-justify-between tw-rounded-md tw-pointer-events-auto tw-cursor-default {{
+    bgColor
+  }}"
+>
+  <div class="tw-flex tw-items-center tw-gap-4 tw-px-2 tw-pb-1 tw-pt-2">
+    <i aria-hidden="true" class="bwi tw-text-xl tw-py-1.5 tw-px-2.5 {{ iconClass }}"></i>
+    <div>
+      <span class="tw-sr-only">{{ variant | i18n }}</span>
+      <div *ngIf="title" class="tw-font-semibold">{{ title }}</div>
```
[the remark this answers is not in the document]
Comment (willmartian) [reply libs/components/src/toast/toast.component.html:10]: I'm unsure what would be better to use here. The [example toasts](https://github.com/scttcper/ngx-toastr/blob/master/src/app/pink.toast.ts#L36-L38) also use `div`. 

I don't _think_ it is valid to use a `h1` etc. 

We could use `p`, but to my knowledge the only a11y advantage of that is SR users can skip from it to the next paragraph, but 1.) that doesn't seem necessary for a short title, and 2.) I don't think paragraphs in aria-live alerts can be skipped.

In short, I don't think choosing something else makes a difference here. @vleague2 what do you think?

File: libs/components/src/toast/toast.component.html [hunk 5]
```diff
@@ -0,0 +1,22 @@
+<div
+  class="tw-mb-1 tw-min-w-[300px] tw-text-contrast tw-flex tw-flex-col tw-justify-between tw-rounded-md tw-pointer-events-auto tw-cursor-default {{
```
[the remark this answers is not in the document]
Comment (willmartian) [reply libs/components/src/toast/toast.component.html]: We only provide a few variants for min-w, none of which are close to this. 

image


We also want to share a single reference to this value between Tailwind and the Toastr css, so I am trying out this token approach: [dfaab89](https://github.com/bitwarden/clients/pull/6490/commits/dfaab8956085f4d4f5f0c3e4c456654d342b0aa6)

File: apps/web/src/app/app.component.ts [hunk 6]
```diff
@@ -205,7 +202,7 @@ export class AppComponent implements OnDestroy, OnInit {
             break;
           }
           case "showToast":
-            this.showToast(message);
+            this.toastService._showToast(message);
```
[the remark this answers is not in the document]
Comment (willmartian) [reply apps/web/src/app/app.component.ts:205]: I want to do that in a separate PR. (Will create a ticket and add to this comment)

I think most calls to `PlatformUtilsService.showToast` already occur in a foreground context.

**Comment by willmartian** [comment]
> @danielleflinn  Looks good! I couldn't get the Service stories to work on my end, but I assume that is a bug with chromatic not storybook.

Yeah, I think it might be some Chromatic bug. Works fine locally.

File: libs/components/src/toast/toast.component.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (vleague2) [reply libs/components/src/toast/toast.component.html:10]: I don't think it affects the user's experience, but I guess a `p` is probably semantically more accurate.

[the remark this answers is not in the document]
Comment (willmartian) [reply libs/components/src/toast/toast.component.html]: Update to use `p` in [eba8678](https://github.com/bitwarden/clients/pull/6490/commits/eba8678015516a9855af6ff3e7a2a0c87880b408)

File: libs/components/src/toast/toast.module.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            default: BitwardenToastrGlobalConfig,
+            config: config,
+          },
+        },
+      ],
+    };
+  }
+}
+
+export const BitwardenToastrGlobalConfig: GlobalConfig = {
```
Comment (vleague2) [inline libs/components/src/toast/toast.module.ts:30]: How would a consumer modify things like auto dismiss for individual toasts? Would we pass it in via the service?

File: libs/angular/src/components/toastr.component.ts [hunk 8]
```diff
@@ -1,98 +0,0 @@
...
-      <i></i>
-    </div>
-    <div>
-      <div *ngIf="title" [class]="options.titleClass" [attr.aria-label]="title">
-        {{ title }} <ng-container *ngIf="duplicatesCount">[{{ duplicatesCount + 1 }}]</ng-container>
-      </div>
-      <div
-        *ngIf="message && options.enableHtml"
-        role="alertdialog"
-        aria-live="polite"
```
[the remark this answers is not in the document]
Comment (vleague2) [reply libs/angular/src/components/toastr.component.ts:35]: So `ngx-toast` already had `aria-live` on it, and we were applying it again here?

**Review by vleague2 — COMMENTED** [review]
Some small questions!

File: libs/components/src/toast/toast.module.ts (hunk 7 above)
Comment (willmartian) [reply libs/components/src/toast/toast.module.ts:30]: Any options that we _want_ to be modifiable at the individual toast level could be added to the `ToastOptions` type and passed in [here.](https://github.com/bitwarden/clients/pull/6490/files/07c5a515af0e5bccef4255e657cae846fb948840#diff-3d71c0041b7d0c25959ecd5ecb4457c587fcb399d7b2fea4375e7260946b2fa3R30-R32) So we could update the service method to allow for the following:

```ts
toastService.showToast({ ... autoDismiss: false });
```

However, one goal of a design system is to limit the amount of control that consumers have in favor of opinionated defaults. _Should_ a consumer be able to alter the auto dismiss behavior, or should it be the same for all toasts?

File: libs/angular/src/components/toastr.component.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (willmartian) [reply libs/angular/src/components/toastr.component.ts:35]: Yes, that is my understand of this https://github.com/scttcper/ngx-toastr/issues/933

File: libs/components/src/toast/toast.module.ts (hunk 7 above)
Comment (danielleflinn) [reply libs/components/src/toast/toast.module.ts:30]: @willmartian If the toast service can calculate a long enough time limit for a user to read the full toast message, than I would be ok with not allowing the consumer alter the dismiss behavior. Right now the rule design uses is to define auto-dismiss behavior for accessibility, is: `a minimum of 5 seconds + 1 extra second per 120 additional words.`

Comment (willmartian) [reply libs/components/src/toast/toast.module.ts:30]: @danielleflinn `autoDismiss` controls whether the toast closes when the timer finishes--I think we always want this to be the case. I believe you are referring to `timeout`, which controls how long the toast timer is. 

> `a minimum of 5 seconds + 1 extra second per 120 additional words`

I can update the service to calculate this and set it as the default timeout. Are there instances where we might not want to follow this rule?

Comment (danielleflinn) [reply libs/components/src/toast/toast.module.ts:30]: I don't think there is any instance where we want the default action to not be close. 

Updating the default timer to calculate this would be great! 

I think there might be instances that may require an override, one that came up recently for me was for the browser passkey flow and adding an excluded domain. It felt too long to keep the toast up for a full 5 seconds before closing the window. I think we opted for 2 or 3 seconds in that instance.

Comment (willmartian) [reply libs/components/src/toast/toast.module.ts:30]: Sounds good! Added the auto timer in [94b337e](https://github.com/bitwarden/clients/pull/6490/commits/94b337e6228797913c6907d6ac02e45379bf8e97)

Comment (vleague2) [reply libs/components/src/toast/toast.module.ts:30]: 👍  I am all for consistent behavior 😎

**Review by vleague2 — DISMISSED** [review]
It looks like all feedback has been addressed except for adding more comments to `toastr.css` I believe, but otherwise looks good.

**Review by justindbaur — DISMISSED** [review]
Looks great, just had a few thoughts about the calculation of the toast timeout but also have no idea how it could be made more locale aware because as far as I know there is no good standardization on "reading speed per character" for locales.

File: libs/components/src/toast/utils.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+ * @param message the toast message to be displayed
+ * @returns the timeout length in milliseconds
+ */
+export const calculateToastTimeout = (message: string | string[]): number => {
+  const paragraphs = Array.isArray(message) ? message : [message];
+  const numWords = paragraphs
+    .map((paragraph) => paragraph.split(/\s+/).filter((word) => word !== ""))
+    .flat().length;
+  return 5000 + Math.floor(numWords / 120) * 1000;
+};
```
[the remark this answers is not in the document]
Comment (willmartian) [reply libs/components/src/toast/utils.ts:14]: @danielleflinn what do you think?

[the remark this answers is not in the document]
Comment (danielleflinn) [reply libs/components/src/toast/utils.ts:14]: I agree with Justin, the limit is likely a bit English-centric but we did pull it form WCAG criteria. I'm not sure how to go about solving it though. Maybe based on the clients selected language we could make the character count dynamic. 

All that said though, toast messages should be pretty short from a design perspective, so cases where we need another 120 characters should be fairly low. Furthermore, Will doesn't this PR include the behavior where if a user mouses over the toast the timer will stop? 

Maybe we leave this as is for now and if we see user feedback about users not having the time to read the full toast, we can iterate on the character count.

[the remark this answers is not in the document]
Comment (willmartian) [reply libs/components/src/toast/utils.ts:14]: > @danielleflinn doesn't this PR include the behavior where if a user mouses over the toast the timer will stop?

Yes, with that in mind, I agree we can wait and see if someone reports the default time as a problem. It seems like something that would be fixable on a case by case level rather than prescriptively here and now. 

---

As for including action/links in toasts: I lean towards toasts being more of an escape hatch for errors. If an error is complex, and benefits from additional user action, we should probably display with something other than a toast in the first place. 

That being said, I am happy to add something like this in another PR if we deem it necessary, but I would want to support it without utilizing raw HTML.

[the remark this answers is not in the document]
Comment (justindbaur) [reply libs/components/src/toast/utils.ts:14]: Awesome, if this number comes from WCAG that is more than satisfactory. I would also hope that very few of our toasts go longer than 120 characters of any locale. Shipping this and keep an ear on the ground works great for me.

@willmartian Very good point, if there is an error we know enough about to write docs about it, then there is likely a much better way to handle that error in app and provide a better experience in that case.

[the remark this answers is not in the document]
Comment (willmartian) [reply libs/components/src/toast/utils.ts:14]: > than 120 characters

Is the rule 120 characters or words? @danielleflinn

[the remark this answers is not in the document]
Comment (justindbaur) [reply libs/components/src/toast/utils.ts:14]: Oh sorry, my brain skipped over the `/\s+/` in the function, you're doing words not characters and the previous discussion was about words. I was the one that brought up characters, the fact that it's words likely mean this works even better in other locales. Sorry.

[the remark this answers is not in the document]
Comment (danielleflinn) [reply libs/components/src/toast/utils.ts:14]: "Words" is what we have specced out in Figma sorry for not catching the mix up earlier

File: libs/components/src/toast/toastr.component.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  `,
+  animations: [
+    trigger("flyInOut", [
+      state("inactive", style({ opacity: 0 })),
+      state("active", style({ opacity: 1 })),
+      state("removed", style({ opacity: 0 })),
+      transition("inactive => active", animate("{{ easeTime }}ms {{ easing }}")),
+      transition("active => removed", animate("{{ easeTime }}ms {{ easing }}")),
+    ]),
+  ],
```
Comment (Hinton) [inline libs/components/src/toast/toastr.component.ts:23]: Would it make more sense for the animation to be in the presentational component? If the goal was to decouple it from the toastr library?

**Review by Hinton — DISMISSED** [review]

File: libs/components/src/toast/toastr.component.ts (hunk 10 above)
Comment (willmartian) [reply libs/components/src/toast/toastr.component.ts:23]: The animation is [applied](https://github.com/scttcper/ngx-toastr/blob/master/src/lib/toastr/toast.component.ts#L64-L67) by the base `Toast` component that this extends from. As such, it made sense to me to keep it part of this wrapper.

**Comment by willmartian** [comment]
I wasn't properly passing the `title` from the service to the component, fixed in [2e30155](https://github.com/bitwarden/clients/pull/6490/commits/2e30155d79fd187ebe786c72d9497117e044ab16).

**Review by vleague2 — DISMISSED** [review]

**Comment by willmartian** [comment]
Latest commits fix the conflicts with https://github.com/bitwarden/clients/pull/8211

**Review by vleague2 — APPROVED** [review]

**Comment by willmartian** [comment]
- Added test ids at QA's request: [ef6be25](https://github.com/bitwarden/clients/pull/6490/commits/ef6be2521262b223399dc8a1cbe0f33a7028f419)
- Fixed a bug on desktop where hovering the toast didn't pause it: [b83d668](https://github.com/bitwarden/clients/pull/6490/commits/b83d668faa390557b81fc1b0fd63eff429bd631f)
- Updated colors to align with recent palette changes: [2a52e6b](https://github.com/bitwarden/clients/pull/6490/commits/2a52e6bb21e329f12f55e8eda998cc1ab2b284b6)

File: libs/components/src/toast/toast.component.html [hunk 11]
```diff
@@ -7,8 +7,10 @@
     <i aria-hidden="true" class="bwi tw-text-xl tw-py-1.5 tw-px-2.5 {{ iconClass }}"></i>
     <div>
       <span class="tw-sr-only">{{ variant | i18n }}</span>
-      <p *ngIf="title" class="tw-font-semibold tw-mb-0">{{ title }}</p>
-      <p *ngFor="let m of messageArray" class="tw-mb-2 last:tw-mb-0">{{ m }}</p>
+      <p *ngIf="title" data-testid="toast-title" class="tw-font-semibold tw-mb-0">{{ title }}</p>
+      <p *ngFor="let m of messageArray" data-testid="toast-message" class="tw-mb-2 last:tw-mb-0">
```
Comment (vleague2) [inline libs/components/src/toast/toast.component.html:11]: Do we need unique test IDs for each message in the array?

Comment (willmartian) [reply libs/components/src/toast/toast.component.html:11]: No, to my understanding QA just needed a query selector to distinguish between title and message. They can query `[data-testid=toast-message]` to get a query list of all of the messages.

**Review by vleague2 — DISMISSED** [review]

**Review by vleague2 — APPROVED** [review]
